entity-service-change-event-generator/src/main/java/org/hypertrace/entity/service/change/event/impl/EntityChangeEventGeneratorImpl.java

```diff
@@ -147,6 +172,67 @@ private void sendUpdateNotification(
     }
   }
 
+  private boolean shouldSendNotification(Entity prevEntity, Entity currEntity) {
+    boolean isAttributesChanged = false, isCoreEntityChanged = false;
+    for (FieldDescriptor fieldDescriptor : Entity.getDescriptor().getFields()) {
+      switch (fieldDescriptor.getNumber()) {
+        case Entity.ATTRIBUTES_FIELD_NUMBER:
+          isAttributesChanged = checkIfAttributesChanged(prevEntity, currEntity);
+          break;
+        default:
+          boolean isFieldChanged =
+              !prevEntity.getField(fieldDescriptor).equals(currEntity.getField(fieldDescriptor));
+          if (!isCoreEntityChanged) {
+            isCoreEntityChanged = isFieldChanged;
+          }
+          break;
+      }
+    }
+
+    return isCoreEntityChanged || isAttributesChanged;
+  }
+
+  private boolean checkIfAttributesChanged(Entity prevEntity, Entity currEntity) {
+    MapDifference<String, AttributeValue> attributesDiff =
+        Maps.difference(prevEntity.getAttributesMap(), currEntity.getAttributesMap());
+    String entityType = prevEntity.getEntityType();
+
+    List<Entry<String, AttributeValue>> addedAttributes =
+        attributesDiff.entriesOnlyOnRight().entrySet().stream()
+            .filter(attributeKey -> !shouldSkipAttribute(entityType, attributeKey.getKey()))
+            .collect(Collectors.toList());
+
+    List<Entry<String, AttributeValue>> deletedAttributes =
+        attributesDiff.entriesOnlyOnLeft().entrySet().stream()
+            .filter(
+                attributeKey ->
+                    !(this.entityToAttributeChangeSkipMap
+                        .get(entityType)
+                        .contains(attributeKey.getKey())))
+            .collect(Collectors.toList());
+
+    List<Entry<String, ValueDifference<AttributeValue>>> updatedAttributes =
+        attributesDiff.entriesDiffering().entrySet().stream()
+            .filter(
+                attributeKey ->
+                    !(this.entityToAttributeChangeSkipMap
+                        .get(entityType)
+                        .contains(attributeKey.getKey())))
+            .collect(Collectors.toList());
+
+    return addedAttributes.size() > 0
+        || deletedAttributes.size() > 0
+        || updatedAttributes.size() > 0;
+  }
+
+  private boolean isCoreEntityChanged(Entity prevEntity, Entity currEntity) {
+    return !prevEntity.getEntityName().equals(currEntity.getEntityName());
+  }
+
+  private boolean shouldSkipAttribute(String entityType, String attributeKey) {
+    return this.entityToAttributeChangeSkipMap.get(entityType).contains(attributeKey);
```

Comment:
this will NPE if it encounters an entity type that didn't happen to be in the config, like the default.